java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowStreamReader.java

```diff
@@ -170,4 +172,21 @@ protected ArrowDictionaryBatch readDictionary() throws IOException {
 
     return MessageSerializer.deserializeDictionaryBatch(result.getMessage(), bodyBuffer);
   }
+
+  @Override
+  protected void readDictionaries(Map<Long, Dictionary> dictionaries) throws IOException {
+    // Read and load all dictionaries from schema
+    for (int i = 0; i < dictionaries.size(); i++) {
+      ArrowDictionaryBatch dictionaryBatch = readDictionary();
+      if (dictionaryBatch == null) {
+        // empty stream has no dictionaries in IPC.
+        if (i == 0) {
+          break;
```

Comment:
If the stream is not empty, and the dictionaries are missing, where will the error be raised? It's a bit hard to test this case I suppose

Reply:
Agreed, I have thought about this and have not good solution. what about check loadNextBatch, like:
if (i == 0) {
          initialized = true;
          if (loadNextBatch()) {
            throw new IOException();
          }
          break;
        }

if first dictionary is null but has follow batch, then i think the dictionaries are missing.

Reply:
We have partially handled this in C++ but it is not as robust as it could be, I just opened

https://issues.apache.org/jira/browse/ARROW-6126

Reply:
Fine, and this PR already updated.

Reply:
Looking at the [spec](https://arrow.apache.org/docs/ipc.html), I think there is a larger bug in the Java code.  It seems to assume that all dictionaries are at the start of the file/stream.  But the specification says:

> DictionaryBatch and RecordBatch messages may be interleaved, but before any dictionary key is used in a RecordBatch it should be defined in a DictionaryBatch

the example however shows all dictionary batches defined and then deltas show afterwards.  Additionally, the java code does not seem to honor the isDelta flag on dictionary batches.

Lastly, we should clarify for the file format, that while:
> In the file format, there is no requirement that dictionary keys should be defined in a DictionaryBatch before they are used in a RecordBatch, as long as the keys are defined somewhere in the file.

That the order of the definition is still important.

@wesm am I reading the spec correctly?   If so I think for this PR we should at least first issue, and we should file follow-up JIRAs for the remaining two.

Reply:
I guess for the first one it seems like the only way you wouldn't need all dictionary batches at the start of the stream is if a dictionary encoded column happens to be all null, so maybe this is a trivial enough edge case.

Reply:
Let me understand, you mean we don't need consider the case "If the stream is not empty, and the dictionaries are missing" ?

Reply:
Suppose we have a record batch of one dictionary encoded column S.  Then according the spec this sequence should be valid:

1.  Schema
2. RecordBatch: S=[null, null, null, null]
3. DictionaryBatch: ['abc', 'efg']
4. Recordbatch: S=[0, 1, 0, 1]

However I think our code would crash at item point 2 (and actually just schema + record batch 1 should be valid).

Reply:
Now the java implementation assumes all dictionaries at the start of the stream, so I think the situation you pointed out is not considered at the moment unless the bug is fixed then? (Maybe not correct, I have limited knowledge about IPC details)

Reply:
Right, I don't think this is fixed.  And I thought it should be fixed here, since it seems like the bug that is being fixed is a special instance of this requirement.

Reply:
I think I understand you, the special case you pointed is right, and I think some more cases are valid too if we assume dictionaries and batches are interleaved:
Have a record batch of two dictionary encoded column S1, S2  below should be valid
1. Schema
2. DictionaryBatch S1: ['ab', 'cd']
3. RecordBatch: S1 = [0,1,0,1] S2 =[null, null, null,]
4. DictionaryBatch S2: ['cc', 'dd']
5. RecordBatch: S1 = [0,1,0,1] S2 =[0,1,0,1]

So what do you think we fix this PR in the following steps and remains write deltas as follow-ups
1. In ArrowStreamReader, do not read all dictionaries at the start
2. When call loadNextBatch, we read message to decide read dictionaries first or directly read a batch, if former, read all dictionaries out before this batch.
3. When we read a batch, we check if the dictionaries it needed has already been read, if not, check if it's all null column and decide whether need throw exception.

In this way, whatever they are interleaved or not, we can parse it properly.

Reply:
@emkornfield I updated this PR according to the solution above.